### DISCUSSION TITLE
Flatpak packager generates icons in template directory

### DISCUSSION THREAD

**Discussion by crschnick** [jreleaser/jreleaser#1336] [Q&A]
Somehow when running the package task, jreleaser will generate a file called `icons/128x128/<dist name>.png` in my template source directory, not the build directory. I.e. the one I set with `templateDirectory`.

**Comment by aalmiray** [comment]
This is by design, as additional files for packagers should come from the packager's `templateDirectory`.

**Reply by crschnick** [reply]
I assumed that jreleaser can take the icon from `project.icon` and use that for the build. Why is it generating that image file in that directory?

**Reply by aalmiray** [reply]
The current implementation expects icon files to exist locally. IIRC `project.icon` may point to a local file instead of a remote URL. https://github.com/jreleaser/jreleaser/blob/b97c7c18f7125ce7e0880f55d189714acafaab7d/core/jreleaser-engine/src/main/java/org/jreleaser/packagers/AppdataUtils.java#L52-L77

**Reply by crschnick** [reply]
But can't the icon file be generated in the build directory then if it needs to exist locally for the build to work?

**Reply by aalmiray** [reply]
Then the packager would need 2 input directories for source files: `templateDirectory` and downloaded assets. At the moment both directories are sourced from the same path. If the downloaded assets directory is realized as a separate entity the question now would be: does its path remain internal (that is, assume `out/jreleaser/downloads/<distribution>/<packager>`) or make it externally configurable?

The internal option would be the straightforward way and requires no changes to the DSL. Yet the pattern should be documented in case someone would like to use the `download` command to setup assets. Moreover, given that there will be 2 file sources, which one wins? I'd suppose the downloaded assets would override any matching assets found in `templateDirectory`.

WDYT?

**Comment by crschnick** [comment]
I guess the most simple solution would just be to use a local project.icon instead of an url.

**Comment by crschnick** [comment]
Well the flatpak topic is kinda done for me now as I found out that it is impossible to launch any other process from a flatpak app via `Runtime.exec()` for example. When I started adding flatpak support I did not expect that there is no sandbox permissions for process spawns. That renders my application essentially useless in flatpak, so I guess I have to look for another packager for Linux.

**Reply by aalmiray** [reply]
I'm afraid you'll face similar issues with `snapcraft` as it also relies on a sandbox. `sdkman` and `brew` may serve as alternatives _but_ they require splitting the portable archive in 2 distributions. Given that xpipe targets 3 platforms with 2 distros each you'll need 6 archives. 

`sdkman` does not handle dependencies between candidates while `brew` does. `sdkman` supports all platforms while `brew` does not support Windows. 

`appImage` might be another option  but I think you may need 2 distros as well.
